refactor(ast): remove `AstKind` for `PropertyKey`

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/11490

This removes the `PropertyKey` AstKind, in support of making the AST kinds more consistent (i.e., only `struct`s should be visited in the AST).

Similar to past PRs, I've added a new `PropertyKeyKind` to narrow the `AstKind` enum (related: https://github.com/oxc-project/backlog/issues/166).

Comments on various changes:
- `id_length` rule: I had to add new handling for nodes which were previously covered by visting `PropertyKey`.
- `no_magic_numbers` rule: Similar to `id_length`, we need to be more careful when checking if a value is part of a class property initializer, since we can't fully rely on the parent kind anymore.
- `prefer_dom_node_text_content` rule: the ancestry here has changed, fortunately for the better in this case. some of the previous juggling of parent/grandparent has been replaced with a simpler check.
- `prefer_string_raw`: the logic for `PropertyKey` was already fully covered by `ObjectProperty`, so I've removed it.
- Semantic binder changes: I added an additional check for binding the `SetAccessor` scope flags which ensures that we only set it when a function is part of a property definition's value, not its key. This evidenced as a test failure in the `no_setter_return` lint rule.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12108?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12108** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12108?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12108?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12105** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12105?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12108).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented:
### Merge activity

* **Jul 8, 10:32 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(ast): remove `AstKind` for `PropertyKey` (#12108)

- part of https://github.com/oxc-project/oxc/issues/11490

This removes the `PropertyKey` AstKind, in support of making the AST kinds more consistent (i.e., only `struct`s should be visited in the AST).

Similar to past PRs, I've added a new `PropertyKeyKind` to narrow the `AstKind` enum (related: https://github.com/oxc-project/backlog/issues/166).

Comments on various changes:
- `id_length` rule: I had to add new handling for nodes which were previously covered by visting `PropertyKey`.
- `no_magic_numbers` rule: Similar to `id_length`, we need to be more careful when checking if a value is part of a class property initializer, since we can't fully rely on the parent kind anymore.
- `prefer_dom_node_text_content` rule: the ancestry here has changed, fortunately for the better in this case. some of the previous juggling of parent/grandparent has been replaced with a simpler check.
- `prefer_string_raw`: the logic for `PropertyKey` was already fully covered by `ObjectProperty`, so I've removed it.
- Semantic binder changes: I added an additional check for binding the `SetAccessor` scope flags which ensures that we only set it when a function is part of a property definition's value, not its key. This evidenced as a test failure in the `no_setter_return` lint rule.
